src/main/scala/org/neo4j/spark/Executor.scala

```diff
@@ -35,79 +35,50 @@ object Executor {
-        if (session.isOpen) {
-          session.close()
-        }
-        driver.close()
-      } catch {
-        case _ => // ignore
-      }
-    }
-
     val driver: Driver = config.driver()
-    val session = driver.session()
+    val session = driver.session(config.sessionConfig())
 
     try {
-      val runner = new TransactionWork[CypherResult]() {
-        override def execute(tx: Transaction): CypherResult = {
-          val result: StatementResult = tx.run(query, toJava(parameters))
-          if (!result.hasNext) {
-            result.consume()
-            session.close()
-            driver.close()
-            return new CypherResult(new StructType(), Iterator.empty)
-          }
-          val peek = result.peek()
-          val keyCount = peek.size()
-          if (keyCount == 0) {
-            val res: CypherResult = new CypherResult(new StructType(), Array.fill[Array[Any]](rows(result))(EMPTY).toIterator)
-            result.consume()
-            close(driver, session)
-            return res
+      val result: Result = session.run(query, toJava(parameters))
+      if (!result.hasNext) {
+        return new CypherResult(new StructType(), Iterator.empty)
+      }
+      val peek = result.peek()
+      val keyCount = peek.size()
+      if (keyCount == 0) {
+        val res: CypherResult = new CypherResult(new StructType(), Array.fill[Array[Any]](rows(result))(EMPTY).toIterator)
+        return res
+      }
+      val keys = peek.keys().asScala
+      val fields = keys.map(k => (k, peek.get(k).`type`())).map(keyType => CypherTypes.field(keyType))
+      val schema = StructType(fields)
+
+      val it = result.asScala.map(record => {
+        val row = new Array[Any](keyCount)
+        var i = 0
+        while (i < keyCount) {
+          val value = record.get(i).asObject() match {
+            case it: util.Map[_, _] => it.asScala.toString()
+            case it: util.Collection[_] => it.toArray()
+            case x => x
           }
-          val keys = peek.keys().asScala
-          val fields = keys.map(k => (k, peek.get(k).`type`())).map(keyType => CypherTypes.field(keyType))
-          val schema = StructType(fields)
-
-          val it = result.asScala.map((record) => {
-            val row = new Array[Any](keyCount)
-            var i = 0
-            while (i < keyCount) {
-              val value = record.get(i).asObject() match {
-                case it: util.Map[_, _] => it.asScala
-                case it: util.Collection[_] => it.toArray()
-                case x => x
-              }
-              row.update(i, value)
-              i = i + 1
-            }
-            if (!result.hasNext) {
-              result.consume()
-              close(driver, session)
-            }
-            row
-          })
-          new CypherResult(schema, it)
+          row.update(i, value)
+          i = i + 1
         }
+        row
+      })
+      new CypherResult(schema, new LazySessionIterator(it, driver, session))
```

Comment:
esp. how does it work to return/leak the driver+session+result out of the spark job processing?